text/0000-input-macros.md

```diff
@@ -0,0 +1,484 @@
+following:
+
+```rs
+let age: i32 = input!()?;
+```
+
+If `input()` didn't trim newlines the above would however always fail since
+the `FromStr` implementations in the standard library don't expect a trailing
+newline. Newline trimming is therefore included for better ergonomics, so that
+programmers don't have to remember to add a `trim` or a `trim_end_matches` call 
+whenever they want to parse the returned string. In cases where newlines have to 
+be preserved the underlying `std::io::Stdin::read_line` can be used directly 
+instead.
+
+This is the default behavior in Python and C# for give you a example, these 
+cases trim newlines by default. In Go, the `bufio.Reader.ReadString()` function
+does not trim newlines, but the `bufio.Scanner` type does. The `bufio.Scanner`
+type is the one that is used in the Go standard library for reading input
+from the console. The `bufio.Scanner` type is a higher level abstraction.
+
+## Why should the function handle EOF?
+
+If the function performs newline trimming, it also has to return an error when
+the input stream reaches EOF because otherwise users had no chance of detecting
+EOF. Handling EOF allows for example a program that attempts to parse each line
+as a number and prints their sum on EOF to be implemented as follows:
+
+```rs
+fn main() -> Result<(), InputError<ParseFloatError>> {
+    let mut sum = 0;
+    loop {
+        let result: Result<f64, _> = input!();
+        match result {
+            Ok(mut prices) => sum += prices,
+            Err(InputError::Eof) => {
+                println!("Oh no, EOF!");
+                break;
+            }
+            Err(e) => {
+                println!("Error: {:?}", e); // e could be a ParseFloatError
+                                            // or an IO error
+                return Err(e);
+            }
+        }
+    }
+    println!("{}", sum);
+    Ok(())
+}
+```
+
+## What is the impact of not doing this?
+If this RFC is not accepted, the current way of reading input from the user
+will remain the same. This means that new Rustaceans will have to learn how to
+use the `std::io::Stdin::read_line` function and how to handle the `Buffer`
+and the `String` types. This can be a barrier to entry for new Rustaceans and
+it can make the language less friendly. 
+
+This RFC was presented as a pre-RFC at a [Rust Argentina meetup](https://rust-lang.ar/p/2025-april/), 
+where it received positive feedback, particularly from attendees new to Rust 
+(many with backgrounds in NodeJS, Python, and Go). They found the current 
+approach to reading input in Rust complex and not very user-friendly, and were 
+enthusiastic about the proposed macros.
+
+They were not too happy with the current way of reading input from the user. 
+They think that it was too complex and not too friendly. 
+
+The presentation can be found [here](https://youtu.be/CjZq93pzOkA?t=4080) (in 
+Spanish)
+And yes, the guy with the black shirt with the Rust logo is me (👋).
```

Comment:
Discussion: Personally, I don't find this argument super convincing. For instance, here are some ecosystem alternatives instead of adding this to the standard library:

(Note that I'm not really expecting you to compare all of these, this listing is just advisory in terms of "what are some of the alternatives are out there?")

- https://github.com/DanielKeep/rust-scan-rules
- https://github.com/anowell/promptly
- https://github.com/jhg/scanf-rs
- https://github.com/wlentz/scan_fmt
- https://github.com/BartMassey/prompted
- https://github.com/oli-obk/rust-si

Notably, observe how the APIs between these ecosystem crates can be quite different, since the crates naturally tailor to different use cases (sometimes overlapping).

Reply:
I think it would be a good idea to test out any proposed API like this as a crate on crates.io first. It's much easier to evolve APIs on crates.io than it is in the standard library, and that way you could get feedback from users on stable rust (rather than asking new learners to install nightly and learn about how to enable unstable features)

And I do think getting feedback from users is helpful for showing that the API is clear and a good fit for the expected usecase.

At least as far as I can tell, your proposed API isn't already published as one of the above-listed crates; if it already is then that would certainly be worth mentioning in the RFC.

Reply:
As related work you can take https://crates.io/crates/text_io which does implement such macros and appears to be used regularly considering ppl open PRs and issues every year